### DISCUSSION TITLE
Ionic React (Mobile) with Spotify OAuth Tokens

### DISCUSSION THREAD

**Discussion by lpmsh** [supabase/supabase#14096] [Questions] [answered]
Hi there,

I'm currently working on an app in which I need to use spotify OAuth within an Ionic React mobile app. 

Currently, my code looks something like this:
```TypeScript
<IonButton
  onClick={async () => {
    const { data, error } = await supabase.auth.signInWithOAuth({
      provider: "spotify",

      options: {
        scopes:
          "streaming user-read-email user-read-private user-read-playback-state user-modify-playback-state",
        skipBrowserRedirect: true,
      },
    });

    if (data.url) {
      openWebPage(data.url);
    }
    await supabase.auth.refreshSession();
  }}
>
  Login
</IonButton>;

```

The `openWebPage` function looks like this: 

```TypeScript
import { InAppBrowser } from "@capgo/inappbrowser";

...


const openWebPage = async (page: string) => {
  await InAppBrowser.open({ url: page });
  await InAppBrowser.setUrl({ url: page });
  InAppBrowser.addListener("urlChangeEvent", async (e) => {
    console.log("CURRENT IN APP BROWSER URL", e.url);

    if (e.url.includes("auxapp://")) {
      const brokenUrl = e.url.split("#");
      const goodURL = brokenUrl.pop();
      console.log("GOOD URL", goodURL);
      const urlParams = new URLSearchParams(goodURL);
      console.log("Url params", JSON.stringify(urlParams));


      //supabase tokens 
      const access_token = urlParams.get("access_token");
      const refresh_token = urlParams.get("refresh_token");

      //example of accessing the spotify tokens from the URL directly
      const spotify_refresh_token = urlParams.get("provider_refresh_token");

      if (access_token && refresh_token) {
        const { data, error } = await supabase.auth.setSession({
          access_token: access_token,
          refresh_token: refresh_token,
        });
      } else {
        console.log("Session setting error");
      }
      await InAppBrowser.close();
    }
  });
};


```

This works fine (although kind of sketchy in my opinion) for just regualr auth to my supabase backend. The problem is that I need to access the spotify access and refresh tokens (preferably through supabase with the `session.provider_token` and `session.provider_refresh_token` properties). As you can see in the openWebPage function, I have an example of accessing the spotify refresh token in the app through the `spotify_refresh_token` variable (The access token is also available through parsing query params), but there is no way for me to set the session's `provider_token` or `provider_refresh_token` through the `setSession` function when I call it. I've been trying to find a better way to do auth for this app as a whole even looking at the swift client for the way that it is done there but I can't seem to figure it out without using all of this weird url parsing stuff. In the regular `signInWithOauth` function the provider token is given to supabase just fine but that is the issue with all of this in app browser redirect nonsense. Preferably I would just have to call that and it would auth the user just like the web. 

Overall, I would say the biggest problem is accessing the spotify tokens through supabase rather than having to mess with storing them myself.

I realize this is a niche use case but I also couldn't find good ways to do this in react native either. 

If someone can point me in the right direction with how to properly use supabase OAuth in my app (with scopes and the external provider tokens) that would be greatly appreciated.

Thanks in advance.

**Comment by GaryAustin1** [comment, marked as answer]
Supabase does not appear to want to store the provider tokens.  You will need to come up with your own approach, or maybe someone has ideas for you.
https://supabase.com/docs/guides/auth/social-login#provider-tokens

_Provider Tokens_

_Once the OAuth flow completes, Supabase Auth will sign your user in. You will receive a copy of the provider token used in the OAuth flow in case you need to use it further. For example, you can use the Google provider token to access Google APIs on behalf of your user._

_Provider tokens are intentionally not stored in your project's database, however. This is because provider tokens give access to potentially sensitive user data in third-party systems. Different applications have different needs, and one application's OAuth scopes may be significantly more permissive than another. If you do want to use the provider token outside of the browser that completed the OAuth flow, you will have to send it manually to a secure server under your control._

**Reply by lpmsh** [reply]
I figured this was the case in terms of storage. Does the way I'm doing the OAuth flow at the moment look right to you? Is there a better way I can be doing it?

**Reply by GaryAustin1** [reply]
I'm not really qualified to comment on ionic or detail using OAuth as I do a javascript PWA and can't use OAuth because redirects don't work well in a PWA.
